refactor: Improve input validation in bootstrap.sh and refactor bit_traits.rs

cypherpepe opened the pull request:
- Improved the input validation logic for the 'clean' command in `bootstrap.sh` to accept various forms of confirmation such as 'y', 'Y', 'yes', and 'YES'.
- Refactored `bit_traits.rs` to reduce code duplication and improve efficiency:
  - Utilized `leading_zeros()` to optimize the `get_msb` function.
  - Implemented a generalized `BitsQueryable` trait for numeric types, reducing the need for multiple implementations.
  - Maintained individual implementations for `FieldElement` and `MemoryAddress` to prevent type-related issues.

cypherpepe committed:
Optimize and refactor bit_traits.rs

- Refactored the BitsQueryable trait to avoid code duplication for numeric types.
- Implemented the get_msb function using the leading_zeros() method for better efficiency.
- Ensured compatibility with u8, u16, u32, u64, u128, and usize through a generalized implementation.
- Maintained individual implementations for FieldElement and MemoryAddress to prevent type-related issues.

cypherpepe committed:
Improve input validation for 'clean' command in bootstrap.sh

- Enhanced the input validation logic for the 'clean' command.
- Now accepts various forms of confirmation such as 'y', 'Y', 'yes', and 'YES'.
- This change makes the script more user-friendly and less prone to input errors.

cypherpepe committed:
Merge branch 'AztecProtocol:master' into patch-1

cypherpepe committed:
Merge branch 'master' into patch-1

fcarreiro reviewed (commented):
Thank you for this! We'll look into it asap. Our CI doesn't work well with external contributors so we need to do a few manual steps.

fcarreiro committed:
Merge branch 'master' into patch-1

fcarreiro commented:
Hi Cypherpepe, this PR does not build: https://github.com/AztecProtocol/aztec-packages/runs/32359062333

```
./avm-transpiler+build *failed* |    Compiling avm-transpiler v0.1.0 (/usr/src/avm-transpiler)
  ./avm-transpiler+build *failed* | error[E0119]: conflicting implementations of trait `BitsQueryable` for type `GenericFieldElement<ark_ff::fields::models::fp::Fp<ark_ff::fields::models::fp::montgomery_backend::MontBackend<ark_bn254::fields::fr::FrConfig, 4>, 4>>`
  ./avm-transpiler+build *failed* |   --> src/bit_traits.rs:21:1
  ./avm-transpiler+build *failed* |    |
  ./avm-transpiler+build *failed* | 15 | impl BitsQueryable for FieldElement {
  ./avm-transpiler+build *failed* |    | ----------------------------------- first implementation here
  ./avm-transpiler+build *failed* | ...
  ./avm-transpiler+build *failed* | 21 | impl<T> BitsQueryable for T
  ./avm-transpiler+build *failed* |    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `GenericFieldElement<ark_ff::fields::models::fp::Fp<ark_ff::fields::models::fp::montgomery_backend::MontBackend<ark_bn254::fields::fr::FrConfig, 4>, 4>>`
  ./avm-transpiler+build *failed* |    |
  ./avm-transpiler+build *failed* |    = note: upstream crates may add a new impl of trait `std::convert::From<acvm::acir::acir_field::GenericFieldElement<ark_ff::fields::models::fp::Fp<ark_ff::fields::models::fp::montgomery_backend::MontBackend<ark_bn254::fields::fr::FrConfig, 4>, 4>>>` for type `u128` in future versions
```

cypherpepe committed:
Optimize bit_traits.rs to fix conflicting trait implementations

- Removed the blanket implementation of the `BitsQueryable` trait for all types `T: Into<u128> + Copy`, which was causing a conflict with the existing implementation for `FieldElement`.
- Added explicit implementations of the `BitsQueryable` trait for primitive integer types (`u8`, `u16`, `u32`, `u64`, `u128`, `usize`) using the `impl_bits_queryable_for_integers!` macro.
- This change resolves the compilation error related to conflicting trait implementations and makes the code more robust against future changes.

cypherpepe commented:
Hi, I've removed the universal implementation and instead provided explicit implementations for the primitive integer types using a macro. This avoids the conflict with `Field Element`.

fcarreiro commented on avm-transpiler/src/bit_traits.rs:
```diff
@@ -65,19 +27,27 @@ impl BitsQueryable for MemoryAddress {
     }
 }
 
+macro_rules! impl_bits_queryable_for_integers {
```
I've consulted with a few people and we think we prefer the explicit BitsQueryable implementations instead of the macro. The rest looks good and still valuable (if small).

Could you make this changes and merge master? Then I can duplicate the PR again, run CI and if it passes we merge.

cypherpepe committed:
Merge branch 'master' into patch-1

cypherpepe committed:
Refactor: explicit BitsQueryable implementations for integer types

cypherpepe commented:
@fcarreiro, I've made the changes you requested:

- Removed the macro and added explicit `BitsQueryable` implementations for all integer types.
- Optimized the `get_msb` function using `leading_zeros()`.
- Updated the branch with the latest changes from `master`.

fcarreiro reviewed (approved):
CI passes in https://github.com/AztecProtocol/aztec-packages/pull/9637
We can merge
